std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+    static if (!XXH_NO_STREAM && XXH_SIZE_OPT >= 2)
+    {
+        /* Simple version, good for code maintenance, but unfortunately slow for small inputs */
+        XXH64_state_t state;
+        XXH64_reset(&state, seed);
+        XXH64_update(&state, cast(const(xxh_u8)*) input, len);
+        return XXH64_digest(&state);
+    }
+    else
+    {
+        if (XXH_FORCE_ALIGN_CHECK)
+        {
+            if (((cast(size_t) input) & 7) == 0)
+            { /* Input is aligned, let's leverage the speed advantage */
+                return XXH64_endian_align(cast(const(xxh_u8)*) input, len,
+                        seed, XXH_alignment.XXH_aligned);
+            }
+        }
+
+        return XXH64_endian_align(cast(const(xxh_u8)*) input, len, seed,
+                XXH_alignment.XXH_unaligned);
+    }
+}
+
+XXH64_state_t* XXH64_createState() @trusted pure nothrow @nogc
+{
+    import core.memory : pureMalloc;
+
+    return cast(XXH64_state_t*) pureMalloc(XXH64_state_t.sizeof);
+
+}
+
+XXH_errorcode XXH64_freeState(XXH64_state_t* statePtr) @trusted pure nothrow @nogc
+{
+    import core.memory : pureFree;
+
+    pureFree(statePtr);
+    return XXH_errorcode.XXH_OK;
+}
+
+void XXH64_copyState(XXH64_state_t* dstState, const XXH64_state_t* srcState) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memcpy;
+
+    memcpy(dstState, srcState, (*dstState).sizeof);
+}
+
+XXH_errorcode XXH64_reset(XXH64_state_t* statePtr, XXH64_hash_t seed) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memset;
+
+    assert(statePtr != null, "statePtr == null");
+    memset(statePtr, 0, (*statePtr).sizeof);
+    statePtr.v[0] = seed + XXH_PRIME64_1 + XXH_PRIME64_2;
+    statePtr.v[1] = seed + XXH_PRIME64_2;
+    statePtr.v[2] = seed + 0;
+    statePtr.v[3] = seed - XXH_PRIME64_1;
+    return XXH_errorcode.XXH_OK;
+}
+
+XXH_errorcode XXH64_update(XXH64_state_t* state, const void* input, size_t len) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memcpy;
+
+    if (input == null)
+    {
+        assert(len == 0, "len != 0");
+        return XXH_errorcode.XXH_OK;
+    }
+
+    {
+        const(xxh_u8)* p = cast(const(xxh_u8)*) input;
+        const xxh_u8* bEnd = p + len;
+
+        state.total_len += len;
+
+        if (state.memsize + len < 32)
+        { /* fill in tmp buffer */
+            memcpy((cast(xxh_u8*) state.mem64) + state.memsize, input, len);
+            state.memsize += cast(xxh_u32) len;
+            return XXH_errorcode.XXH_OK;
+        }
+
+        if (state.memsize)
+        { /* tmp buffer is full */
+            memcpy((cast(xxh_u8*) state.mem64) + state.memsize, input, 32 - state.memsize);
+            state.v[0] = XXH64_round(state.v[0], XXH_readLE64(&state.mem64[0]));
+            state.v[1] = XXH64_round(state.v[1], XXH_readLE64(&state.mem64[1]));
+            state.v[2] = XXH64_round(state.v[2], XXH_readLE64(&state.mem64[2]));
+            state.v[3] = XXH64_round(state.v[3], XXH_readLE64(&state.mem64[3]));
```

Comment:
This is one of the things we can later vectorize. Probably not doing it on this PR, but worth leaving a comment?

Reply:
I think, such optimisations should be in a later PR. First, the code should work using generic code, then we can move to special things.